### PR TITLE
Fault tolerance for SE

### LINKED ISSUE #2110: Create fault tolerance API and implementation for Helidon SE
## Environment Details
* Helidon Version: 2.0+
* Helidon SE

----------

## Problem Description
Create a fault tolerance usable in Helidon SE (e.g. fully supporting reactive execution). Provide similar features to MP Fault Tolerance.

* Async - should be designed to handle any blocking operation in Helidon SE
* Circuit breaker
* Bulkhead
* Fallback
* Retry
* Timeout

Related to #1194

### PR DESCRIPTION
**Pull request by tomas-langer**
Fault tolerance handlers supporting `Single` and `Multi`.

Resolves #2110 
Using `ServerResponse.send(Throwable)` now resolves error handlers and gets the cause of `CompletionException` to support `CompletionStage.exceptionally`

### CODE REVIEW

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        AtomicInteger count = new AtomicInteger();
+
+        TestSubscriber ts = new TestSubscriber();
+
+        Multi<Integer> multi = retry.invokeMulti(() -> {
+            if (count.getAndIncrement() == 0) {
+                //return new PartialPublisher();
+                return Multi.concat(Multi.just(0), Multi.error(new RetryException()));
+            } else {
+                TestSubscriber it = ts;
```
Comment (spericas) [inline fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java:211]: Need it?

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/BulkheadTest.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    @Test
+    void testBulkheadQueue() throws InterruptedException {
+        Bulkhead bulkhead = Bulkhead.builder()
+                .limit(1)
+                .queueLength(1000)
+                .build();
+
+        Request inProgress = new Request(0);
+        Single<Integer> inProgressResult = bulkhead.invoke(inProgress::invoke);
```
Comment (spericas) [inline fault-tolerance/src/test/java/io/helidon/faulttolerance/BulkheadTest.java:57]: Variable `inProgressResult` not used

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/FaultToleranceTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        CompletionException completionException = assertThrows(CompletionException.class,
+                                                               () -> result.await(1, TimeUnit.SECONDS));
+        Throwable cause = completionException.getCause();
+        assertThat(cause, notNullValue());
+        assertThat(cause, instanceOf(expected));
+
+        return expected.cast(cause);
+    }
+
+    private class Manual {
```
Comment (spericas) [inline fault-tolerance/src/test/java/io/helidon/faulttolerance/FaultToleranceTest.java:100]: Minor: can be static

File: examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+import io.helidon.faulttolerance.Timeout;
+import io.helidon.webserver.Routing;
+import io.helidon.webserver.ServerRequest;
+import io.helidon.webserver.ServerResponse;
+import io.helidon.webserver.Service;
+
+/**
+ * Simple service to demonstrate fault tolerance.
+ */
+public class FtService implements Service {
```
Comment (spericas) [inline examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java:37]: May be a bit more readable if methods had a suffix like "M", a bit difficult to read when named same as variables

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/AsyncTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+        CompletionException exception = assertThrows(CompletionException.class, () -> result.await(1, TimeUnit.SECONDS));
+
+        Throwable cause = exception.getCause();
+
+        assertThat(cause, notNullValue());
+        assertThat(cause, instanceOf(MyException.class));
+    }
+
+    private List<String> syncList() {
```
Comment (spericas) [inline fault-tolerance/src/test/java/io/helidon/faulttolerance/AsyncTest.java:67]: Unused

File: fault-tolerance/src/main/java/io/helidon/faulttolerance/AtomicCycle.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    int incrementAndGet() {
+        int currentIndex;
+        int nextIndex;
+        do {
+            currentIndex = atomicInteger.get();
+            nextIndex = (currentIndex == maxIndex) ? 0 : currentIndex + 1;
+        } while (!atomicInteger.compareAndSet(currentIndex, nextIndex));
+
+        return nextIndex;
+    }
```
Comment (spericas) [inline fault-tolerance/src/main/java/io/helidon/faulttolerance/AtomicCycle.java:38]: Isn't this better?

```
    final static class AtomicCycle {
        private final AtomicInteger atomicInteger = new AtomicInteger(-1);
        private final int maxIndex;

        AtomicCycle(int maxIndex) {
            this.maxIndex = maxIndex + 1;
        }

        int incrementAndGet() {
            return atomicInteger.accumulateAndGet(maxIndex, (current, max) -> (current + 1) % max);
        }
    }
```

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java (hunk 1 above)
Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java:211]: Need what? Sorry I do not see it

File: examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java (hunk 4 above)
Comment (tomas-langer) [reply examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java:37]: Will make it clear

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java (hunk 1 above)
Comment (spericas) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java:211]: :) Meant if we need the variable `it`

Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java:211]: No :)

File: examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java (hunk 4 above)
Comment (tomas-langer) [reply examples/webserver/fault-tolerance/src/main/java/io/helidon/webserver/examples/faulttolerance/FtService.java:37]: Changed method names to `*Handler`

File: fault-tolerance/src/main/java/io/helidon/faulttolerance/AtomicCycle.java (hunk 6 above)
Comment (tomas-langer) [reply fault-tolerance/src/main/java/io/helidon/faulttolerance/AtomicCycle.java:31]: Yes, fixed.

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/AsyncTest.java (hunk 5 above)
Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/AsyncTest.java]: Removed

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/FaultToleranceTest.java (hunk 3 above)
Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/FaultToleranceTest.java]: Static

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/BulkheadTest.java (hunk 2 above)
Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/BulkheadTest.java]: Removed

File: fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java (hunk 1 above)
Comment (tomas-langer) [reply fault-tolerance/src/test/java/io/helidon/faulttolerance/RetryTest.java]: it not there anymore

**Pull request merged**